internal/engine/build_and_deployer_test.go

```diff
@@ -51,6 +55,88 @@ func TestDockerForMacDeploy(t *testing.T) {
 	}
 }
 
+func TestIncrementalBuild(t *testing.T) {
+	f := newBDFixture(t, k8s.EnvDockerDesktop)
+	defer f.TearDown()
+
+	_, err := f.bd.BuildAndDeploy(f.Ctx(), SanchoService, NewBuildState(alreadyBuilt))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if f.docker.PushCount != 0 {
+		t.Errorf("Expected no push to docker, actual: %d", f.docker.PushCount)
+	}
+	if f.docker.CopyCount != 1 {
+		t.Errorf("Expected 1 copy to docker container call, actual: %d", f.docker.PushCount)
+	}
+	if len(f.docker.ExecCalls) != 1 {
+		t.Errorf("Expected 1 exec in container call, actual: %d", len(f.docker.ExecCalls))
+	}
+	if len(f.docker.RestartsByContainer) != 1 {
+		t.Errorf("Expected 1 container to be restarted, actual: %d", len(f.docker.RestartsByContainer))
+	}
+}
+
+func TestFallBackToImageDeploy(t *testing.T) {
+	f := newBDFixture(t, k8s.EnvDockerDesktop)
+	defer f.TearDown()
+
+	nt, err := k8s.ParseNamedTagged("gcr.io/some-project-162817/sancho:foo")
+	if err != nil {
+		t.Fatal(err)
+	}
+	br := BuildResult{
+		Image: nt,
+	}
+
+	newBR, err := f.bd.BuildAndDeploy(f.Ctx(), SanchoService, NewBuildState(br))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if f.docker.PushCount != 0 {
```

Comment:
maybe worth doing this with GKE so we DO expect a push to docker and extra obvious if we don't get one? cuz we use the same assertion (`pushCount != 0`) below to assert that we used container-based build...)

Reply:
I think we're already doing this?

https://github.com/windmilleng/tilt/blob/71233b012dcc87b443023e32989b6bf09936a22c/internal/engine/build_and_deployer_test.go#L29